core/src/unprocessed_packet_batches.rs

```diff
@@ -0,0 +1,1205 @@
+                is_simple_vote,
+            })
+        } else {
+            None
+        }
+    }
+
+    /// Read the transaction message from packet data
+    fn packet_message(packet: &Packet) -> Option<&[u8]> {
+        let (sig_len, sig_size) = decode_shortu16_len(&packet.data).ok()?;
+        let msg_start = sig_len
+            .checked_mul(size_of::<Signature>())
+            .and_then(|v| v.checked_add(sig_size))?;
+        let msg_end = packet.meta.size;
+        Some(&packet.data[msg_start..msg_end])
+    }
+
+    // Returns whether the given `PacketBatch` has any more remaining unprocessed
+    // transactions
+    pub fn update_buffered_packets_with_new_unprocessed(
+        &mut self,
+        _original_unprocessed_indexes: &[usize],
+        new_unprocessed_indexes: &[usize],
+    ) -> bool {
+        let has_more_unprocessed_transactions = !new_unprocessed_indexes.is_empty();
+        if has_more_unprocessed_transactions {
+            self.unprocessed_packets
+                .retain(|index, _| new_unprocessed_indexes.contains(index));
+        } else {
+            self.unprocessed_packets.clear();
+        }
+
+        has_more_unprocessed_transactions
+    }
+}
+
+// TODO TAO - refactor type into struct
+pub type UnprocessedPacketBatches = VecDeque<DeserializedPacketBatch>;
+
+// prioritize unprocessed packets in buffered packet_batches by its fee/CU then by its sender's
+// stakes
+pub fn prioritize_by_fee_then_stakes(
+    unprocessed_packet_batches: &UnprocessedPacketBatches,
+    working_bank: Option<Arc<Bank>>,
+    packet_sender_info: &mut Option<PacketSenderInfo>,
+) -> Vec<PacketLocator> {
+    let (stakes, locators) =
+        get_stakes_and_locators(unprocessed_packet_batches, packet_sender_info);
+
+    // 2. weight shuffle -> shuffled locators
+    let shuffled_packet_locators = weighted_shuffle(&stakes, &locators, packet_sender_info);
+
+    // 3. sort by fee function -> sorted and shuffled locators
+    prioritize_by_fee(
+        unprocessed_packet_batches,
+        &shuffled_packet_locators,
+        working_bank,
+    )
+}
+
+// Iterates packets in buffered batches, returns all unprocessed packet's stake,
+// and its location (batch_index plus packet_index within batch)
+fn get_stakes_and_locators(
+    unprocessed_packet_batches: &UnprocessedPacketBatches,
+    packet_sender_info: &mut Option<PacketSenderInfo>,
+) -> (Vec<u64>, Vec<PacketLocator>) {
+    let mut stakes = Vec::<u64>::with_capacity(TOTAL_BUFFERED_PACKETS);
+    let mut locators = Vec::<PacketLocator>::with_capacity(TOTAL_BUFFERED_PACKETS);
+
+    unprocessed_packet_batches.iter().enumerate().for_each(
+        |(batch_index, deserialized_packet_batch)| {
+            let packet_batch = &deserialized_packet_batch.packet_batch;
+            deserialized_packet_batch
+                .unprocessed_packets
+                .keys()
+                .for_each(|packet_index| {
+                    let p = &packet_batch.packets[*packet_index];
```

Comment:
This pattern seems to come up a lot, let's add a new function to `UnprocessedPacketBatches`:

```
    pub fn packet_iter(&self) -> impl Iterator<Item = &Packet> {
        self.unprocessed_packets
            .keys()
            .map(|index| &self.packet_batch.packets[*index])
    }
```